src/core/RandomVariables.jl

```diff
@@ -83,97 +168,514 @@ function Turing.runmodel!(model::Model, vi::AbstractVarInfo, spl::AbstractSample
+    # Each random variable can be sampled using multiple samplers, e.g. in Gibbs, hence the `Set`
+    gids        ::    TGIds # AbstractVector{Set{Selector}}
+
+    # Number of `observe` statements before each random variable is sampled
+    orders      ::    Vector{Int}
+
+    # Each `flag` has a `BitVector` `flags[flag]`, where `flags[flag][i]` is the true/false flag value corresonding to `vns[i]`
     flags       ::    Dict{String, BitVector}
+end
+
+"""
+`Metadata()`
+
+Constructs an empty type unstable instance of `Metadata`.
+"""
+function Metadata()
+    vals  = Vector{Real}()
+    flags = Dict{String, BitVector}()
+    flags["del"] = BitVector()
+    flags["trans"] = BitVector()
+
+    return Metadata(
+        Dict{VarName, Int}(),
+        Vector{VarName}(),
+        Vector{UnitRange{Int}}(),
+        vals,
+        Vector{Distributions.Distribution}(),
+        Vector{Set{Selector}}(),
+        Vector{Int}(),
+        flags
+    )
+end
 
-    UntypedVarInfo() = begin
-        vals  = Vector{Real}()
-        logp  = 0.0
-        flags = Dict{String, BitVector}()
-        flags["del"] = BitVector()
-        flags["trans"] = BitVector()
-
-        new(
-            Dict{VarName, Int}(),
-            Vector{VarName}(),
-            Vector{UnitRange{Int}}(),
-            vals,
-            Vector{Distributions.Distribution}(),
-            Vector{Set{Selector}}(),
-            logp,
-            0,
-            Vector{Int}(),
-            flags
-        )
+"""
+`empty!(meta::Metadata)`
+
+Empties all the fields of `meta`. This is useful when using a sampling algorithm that assumes an empty `meta`, e.g. `SMC`. 
+"""
+function Base.empty!(meta::Metadata)
+    empty!(meta.idcs)
+    empty!(meta.vns)
+    empty!(meta.ranges)
+    empty!(meta.vals)
+    empty!(meta.dists)
+    empty!(meta.gids)
+    empty!(meta.orders)
+    for k in keys(meta.flags)
+        empty!(meta.flags[k])
     end
+
+    return meta
 end
-VarInfo() = UntypedVarInfo()
 
-const VarView = Union{Int,UnitRange,Vector{Int},Vector{UnitRange}}
+###########
+# VarInfo #
+###########
+
+"""
+```
+struct VarInfo{Tmeta, Tlogp} <: AbstractVarInfo
+    metadata::Tmeta
+    logp::Base.RefValue{Tlogp}
+    num_produce::Base.RefValue{Int}
+end
+```
 
+A light wrapper over one or more instances of `Metadata`. Let `vi` be an instance of `Metadata`. If `vi isa VarInfo{<:Metadata}`, then only `Metadata` instance is used for all the sybmols. `VarInfo{<:Metadata}` is aliased `UntypedVarInfo`. If `vi isa VarInfo{<:NamedTuple}`, then `vi.metadata` is a `NamedTuple` that maps each symbol used on the LHS of `~` in the model to its `Metadata` instance. The latter allows for the type specialization of `vi` after the first sampling iteration when all the symbols have been observed. `VarInfo{<:NamedTuple}` is aliased `TypedVarInfo`.
+
+Note: It is the user's responsibility to ensure that each symbol is visited at least once whenever the model is called, regardless of any stochastic branching.
```

Comment:
presumably, it's ok if a variable is visited multiple times since they can have different `counter`? If so, can we say this explicitly in the comment to reduce potential confusion?

Reply:
Each random variable can only be visited once in a model call. But the symbol of the Julia variable can be visited more than once, e.g. `x[1] ~ ...` and `x[2] ~ ...`. This line refers to the symbol `x` not the random variables `x[1]` and `x[2]`. The requirement here is that each symbol, e.g. `x`, is visited at least once. I will try to make this clearer.